src/solvers/refinement/string_refinement.cpp

```diff
@@ -1004,26 +1037,70 @@ exprt fill_in_array_with_expr(
 {
   PRECONDITION(expr.type().id()==ID_array);
   PRECONDITION(expr.id()==ID_with || expr.id()==ID_array_of);
-
+  const array_typet &array_type=to_array_type(expr.type());
+#if 0
   // Nothing to do for empty array
   if(expr.id()==ID_array_of)
     return expr;
+#endif
 
   // Map of the parts of the array that are initialized
   std::map<std::size_t, exprt> initial_map;
 
+  // Set the last index to be sure the array will have the right length
+  const auto array_size_opt=expr_cast<std::size_t>(array_type.size());
+  if(array_size_opt && *array_size_opt>0)
+    initial_map.emplace(
+      *array_size_opt-1,
+      from_integer(CHARACTER_FOR_UNKNOWN, array_type.subtype()));
+
   for(exprt it=expr; it.id()==ID_with; it=to_with_expr(it).old())
   {
     // Add to `initial_map` all the pairs (index,value) contained in `WITH`
     // statements
     const with_exprt with_expr=to_with_expr(it);
     const exprt &then_expr=with_expr.new_value();
     const auto index=expr_cast_v<std::size_t>(with_expr.where());
-    if(index<string_max_length)
+    if(index<string_max_length && (!array_size_opt || index<*array_size_opt))
       initial_map.emplace(index, then_expr);
   }
 
-  array_exprt result(to_array_type(expr.type()));
+  array_exprt result(array_type);
+  result.operands()=fill_in_map_as_vector(initial_map);
+  return result;
+}
+
+/// Fill an array represented by an array_expr by propagating values to
+/// the left for unknown values. For instance `{ 24 , * , * , 42, * }` will give
+/// `{ 24, 42, 42, 42, '?' }`
+/// \param expr: an array expression
+/// \param string_max_length: bound on the length of strings
+/// \return an array expression with filled in values
+exprt fill_in_array_expr(const array_exprt &expr, std::size_t string_max_length)
+{
+  PRECONDITION(expr.type().id()==ID_array);
+  array_typet array_type=to_array_type(expr.type());
+  PRECONDITION(array_type.subtype().id()==ID_unsignedbv);
+
+  // Map of the parts of the array that are initialized
+  std::map<std::size_t, exprt> initial_map;
+  const auto array_size_opt=expr_cast<std::size_t>(array_type.size());
+
+  if(array_size_opt && *array_size_opt>0)
+    initial_map.emplace(
+      *array_size_opt-1,
+      from_integer(CHARACTER_FOR_UNKNOWN, array_type.subtype()));
+
+  for(std::size_t i=0; i<expr.operands().size(); ++i)
+  {
+    if(i<string_max_length && expr.operands()[i].id()!=ID_unknown)
+      initial_map[i]=expr.operands()[i];
+    // Special case for unknown last character
+    if(i==expr.operands().size()-1 && expr.operands()[i].id()==ID_unknown)
```

Comment:
What is this case for, what does it achieve?

Reply:
The "what" can be deduced from the code. As such the comment should be removed. @romainbrenguier should comment on why there is a special case for the last character (if that's not obvious).

Reply:
@romainbrenguier note to self: This overrides the if above.